### PR TITLE
Buffer views (first part of zero copy offset buffers)

### PR DESCRIPTION
**Pull request by uuuvn**
https://github.com/tinygrad/tinygrad/pull/3924#issuecomment-2023126645

### CODE REVIEW

**Comment by uuuvn** [comment]
Also right now when it uncows buffer it copies it to cpu and back, i will fix that tomorrow

**Comment by geohot** [comment]
Why does it have copy on write? I would expect it to actually just be the same memory.

File: tinygrad/device.py [hunk 1]
```diff
@@ -129,6 +129,7 @@ def _alloc(self, size:int, options:BufferOptions): return (ctypes.c_uint8 * size
   def as_buffer(self, src) -> memoryview: return flat_mv(memoryview(src))
   def copyin(self, dest, src:memoryview): ctypes.memmove(dest, from_mv(src), len(src))
   def copyout(self, dest:memoryview, src): ctypes.memmove(from_mv(dest), src, len(dest))
+  def offset(self, buf, offset: int, size:int): return ctypes.cast(ctypes.byref(buf, offset), ctypes.POINTER(ctypes.c_uint8 * size)).contents
```
Comment (geohot) [inline tinygrad/device.py:132]: I think this function signature works.

**Comment by geohot** [comment]
In fact, it definitely shouldn't be copy on write. Think about the disk backend.

**Comment by uuuvn** [comment]
The copy on write is optional. The idea is that by default it behaves like a copy (no write passthrough) to save memory without everything above the Buffer level except scheduler (which creates those CoW copies) even knowing about it.

The cow=False feature is for setitem (including DISK one) to work properly. Which would probably be just a flag to contiguous

If i make it always cow=False which is like very easy and i had it a long time ago and most of the time was actually spent on CoW stuff and testing it then on backends that support offsets something like:
```python
b = a.shrink(((4,16),)).contiguous()
b.assign(c)
print(a.numpy())
```
would give different result than on a backend that doesn't support it. Also it might break things that rely on this behaviour.

**Comment by uuuvn** [comment]
Read only is probably fine but then it's not really possible to know if something is going to be written to and so when something from eg another run_schedule attempts to write into it we just end up with either incorrect behaviour or just doing CoW at another level

**Comment by uuuvn** [comment]
Also see the `test_simple_onelevel` test where both cow=False behaviour (for DISK and setitem) and cow=True (for saving memory when .contiguous() can be done copy-free

**Comment by uuuvn** [comment]
I mean if you think that cow=True stuff is useless i can just make it always passthrough but then it's a potential for bugs in future when we assign to something that looks like it should be just itself but actually some other tensor is modified because the thing we assigned to was a offset pseudo-copy of that other thing we don't expect to change

**Comment by uuuvn** [comment]
maybe i'm just overthinking stuff idk

**Comment by uuuvn** [comment]
> Why does it have copy on write? I would expect it to actually just be the same memory (or maybe read only).

#4008 is the same as this pr except always cow=False , ready to review

**Comment by uuuvn** [comment]
closing in favor of #4008

**Pull request closed without merging**